lib/actions/user.actions.ts

```diff
@@ -0,0 +1,56 @@
+'use server';
+
+import { prisma } from '@/db'; //prisma db connection
+
+export async function getUser(email: string) {
+  try {
+    const user = prisma.user.findUnique({
+      where: {
+        email,
+      },
+    });
+    return user;
+  } catch (error) {
+    console.error('Error finding user:', error);
+    throw new Error('User not found!');
+  }
+}
+
+export async function updateUser(data: {
+  id: string;
+  name: string;
+  email: string;
+}) {
+  try {
+    if (data) {
+      const user = prisma.user.update({
+        where: {
+          id: data.id,
+        },
+        data: {
+          ...data,
+          name: data.name,
+          email: data.email,
+        },
+      });
+      return user;
+    }
+  } catch (error) {
+    console.error('Error updating user:', error);
+    throw new Error('An unexpected error occurred while updating user.');
+  }
+}
```

Comment:
For the update, you're requiring the email, name, and the id to run an update. What if we just wanted to update the name? This could be tedious, and then for every other field you need to explicitly add each option ('role' for example).

An easier way, while still giving you type safety, would be to use the typescript "Partial" in order to tell your environment that we expect at least "some" of a user.

We can then pass 'data' in, and we'll handle updating all the fields.

```js
export async function updateUser(data: Partial<User>) {
  try {
    if (data) {
      const user = prisma.user.update({
        where: {
          id: data.id,
        },
        data,
      });
      return user;
    }
  } catch (error) {
    console.error("Error updating user:", error);
    throw new Error("An unexpected error occurred while updating user.");
  }
}
```
Now you can update a userbyID without adding to this function every time you need to add more fields to access or update.

Plus get really great typesafety!


https://github.com/DDVVPP/GitNote/assets/4108484/c2a36352-e0f8-4be1-9083-7f765fc66a14

Reply:
Great, thank you!

Question about the data - Even though I'm sending the data as a whole object, it doesn't replace all of the data fields in the db?  It knows to replace only the selected fields that I've sent?